### PR TITLE
[Dynamic Control] Phase 4: Add ordered, coalesced PolicyStore subscriptions

### PR DESCRIPTION
**Pull request by stevejgordon**
Contributes to: #4742

## Changes

Adds PolicyChangeNotifier/PolicyChangeSubscription so that a subscriber can register with PolicyStore.Subscribe and receive an atomic replay of the current snapshot followed by every subsequent commit, coalesced to the latest revision and never delivered out of order or overlapping. Subscriber callback failures are isolated per-subscription and reported via a new DynamicControlEventSource rather than propagating to the committing thread.

PolicyStore is now IDisposable to release its subscriptions.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] ~Changes in public API reviewed (if applicable)~

### CODE REVIEW

File: src/OpenTelemetry.DynamicControl/Internal/Diagnostics/DynamicControlEventSource.cs
Comment (martincostello) [inline src/OpenTelemetry.DynamicControl/Internal/Diagnostics/DynamicControlEventSource.cs:1]: Please add a test for this [like this](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/test/OpenTelemetry.Extensions.Tests/EventSourceTests.cs).

File: src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+        lock (this.registrationLock)
+        {
+            this.ThrowIfDisposed();
+
+            var subscription = new PolicyChangeSubscription(onChanged, this);
+            var current = this.subscribers;
+            var updated = new PolicyChangeSubscription[current.Length + 1];
+            Array.Copy(current, updated, current.Length);
+            updated[current.Length] = subscription;
```
Comment (martincostello) [inline src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs]: I think this would be _slightly_ more readable as `updated[updated.Length - 1]`, as otherwise you need to read more than one line to check this is correctly adding to the end of the array.

It's probably also a stronger hint to the JIT that the write is in bounds.

File: src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            Volatile.Write(ref this.subscribers, updated);
+
+            return subscription;
+        }
+    }
+
+    /// <summary>
+    /// Removes a subscription. Called by <see cref="PolicyChangeSubscription.Dispose"/>.
```
Comment (martincostello) [inline src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs]: One less thing to update if that ever changes.

```suggestion
    /// Removes a subscription.
```

File: src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            if (current.Length == 1)
+            {
+                Volatile.Write(ref this.subscribers, EmptySubscribers);
+                return;
+            }
+
+            var updated = new PolicyChangeSubscription[current.Length - 1];
+            Array.Copy(current, updated, index);
+            Array.Copy(current, index + 1, updated, index, current.Length - index - 1);
+            Volatile.Write(ref this.subscribers, updated);
```
Comment (martincostello) [inline src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs:103]: As I've spent a minute trying to trace the bounds arithmetic in my head here, if not already present let's make sure there's test coverage for these boundary cases:

- 2 & 3 element array, remove the first item
- 2 & 3 element array, remove the last item
- 3 element array, remove the middle item

File: src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeSubscription.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+            this.dispatching = true;
+        }
+
+        ThreadPool.UnsafeQueueUserWorkItem(static state => ((PolicyChangeSubscription)state!).Drain(), this);
+    }
+
+    private void Drain()
+    {
+        while (true)
```
Comment (martincostello) [inline src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeSubscription.cs:94]: 1. Should we have a `CancellationTokenSource` to signal from `Dispose` to stop the loop here (I'm always extra wary of `while (true)`).
2. Should we have any backpressure on how often this publishes a change?

File: test/OpenTelemetry.DynamicControl.Tests/PolicyStoreConcurrencyTests.cs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        while (!condition())
+        {
+            if (sw.ElapsedMilliseconds > timeoutMilliseconds)
+            {
+                Assert.Fail("Timed out waiting for condition.");
+            }
+
+            await Task.Delay(10);
+        }
+    }
```
Comment (martincostello) [inline test/OpenTelemetry.DynamicControl.Tests/PolicyStoreConcurrencyTests.cs]: Move to a helper as there's ~~at least two~~ three copies of the same pattern here?

File: src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeSubscription.cs (hunk 4 above)
Comment (stevejgordon) [reply src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeSubscription.cs:94]: On the CTS: I don't think we need one here. The exit condition (`disposed || pending is null`) is checked under the same lock that `Dispose()` sets `disposed` under, so the loop is guaranteed to observe disposal and exit on the next itteration.

On backpressure: I had AI reassess against the telemetry-policy OTEP and it requires policy updates to be async and fail-open, and explicitly says a stale/out-of-date policy must never be lethal to the system. Blocking the committer to apply backpressure would violate that, so coalesce-to-latest (current behavior) is spec-aligned, not a gap. I'll add a comment citing this so it reads as intentional rather than an oversight.

File: test/OpenTelemetry.DynamicControl.Tests/PolicyStoreConcurrencyTests.cs (hunk 5 above)
Comment (stevejgordon) [reply test/OpenTelemetry.DynamicControl.Tests/PolicyStoreConcurrencyTests.cs]: Addressed in 2c875ac5eeb2edf29e77d935986cce92ffe9434a

File: src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs (hunk 3 above)
Comment (stevejgordon) [reply src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs:103]: Addressed in 2c875ac5eeb2edf29e77d935986cce92ffe9434a

File: src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs (hunk 2 above)
Comment (stevejgordon) [reply src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs]: Addressed in 2c875ac5eeb2edf29e77d935986cce92ffe9434a

File: src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs (hunk 1 above)
Comment (stevejgordon) [reply src/OpenTelemetry.DynamicControl/Internal/Store/PolicyChangeNotifier.cs]: Addressed in 2c875ac5eeb2edf29e77d935986cce92ffe9434a

File: src/OpenTelemetry.DynamicControl/Internal/Diagnostics/DynamicControlEventSource.cs
Comment (stevejgordon) [reply src/OpenTelemetry.DynamicControl/Internal/Diagnostics/DynamicControlEventSource.cs:1]: Addressed in 2c875ac5eeb2edf29e77d935986cce92ffe9434a

File: test/OpenTelemetry.DynamicControl.Tests/OpenTelemetry.DynamicControl.Tests.csproj [hunk 6]
```diff
@@ -12,6 +12,7 @@
 
   <ItemGroup>
     <Compile Include="$(RepoRoot)\test\Shared\InMemoryEventListener.cs" Link="Includes\InMemoryEventListener.cs" />
+    <Compile Include="$(RepoRoot)\test\Shared\EventSourceTestHelper.cs" Link="Includes\EventSourceTestHelper.cs" />
```
Comment (martincostello) [inline test/OpenTelemetry.DynamicControl.Tests/OpenTelemetry.DynamicControl.Tests.csproj:14]: Nit: sorting.

File: test/OpenTelemetry.DynamicControl.Tests/PolicyStoreConcurrencyTests.cs (hunk 5 above)
Comment (martincostello) [reply test/OpenTelemetry.DynamicControl.Tests/PolicyStoreConcurrencyTests.cs]: Looks like this didn't get committed.

**Comment by stevejgordon** [comment]
> Looks like this didn't get committed.

I'm confused? It's called from the three places?

**Comment by stevejgordon** [comment]
> I'm confused? It's called from the three places?

Oh wait. I see. I fixed a slightly different issue!

**Comment by martincostello** [comment]
> I'm confused? It's called from the three places?

- `PolicyChangeSubscriptionTests.WaitUntil()`
- `PolicyStoreConcurrencyTests.WaitUntil()`
- `PolicyStoreSubscriptionTests.WaitUntil()`

These three methods all seem identical except for having a different default timeout and delay.

**Review by martincostello — APPROVED** [review]
